### PR TITLE
Update external api to use pointers

### PR DESCRIPTION
**Pull request by kargakis**
Part of AZURE-135

Fixes https://github.com/openshift/openshift-azure/issues/434

### CODE REVIEW

**Comment by kargakis** [comment]
This should be ready for a review, we could also try sneaking it in so I will avoid rebase hell and I am going to work on more unit tests in a follow-up.

File: pkg/api/converterfromv20180930preview.go [hunk 1]
```diff
@@ -6,92 +6,314 @@ import (
...
+		// provider.
+		if index == -1 {
+			cs.Properties.AuthProfile.IdentityProviders = append(cs.Properties.AuthProfile.IdentityProviders, convertIdentityProvider(ip))
+		} else {
+			provider := cs.Properties.AuthProfile.IdentityProviders[index].Provider
+			switch out := provider.(type) {
+			case (*AADIdentityProvider):
+				in := ip.Provider.(*v20180930preview.AADIdentityProvider)
+				if in.Kind != nil {
+					out.Kind = *in.Kind
```
Comment (mjudeikis) [inline pkg/api/converterfromv20180930preview.go]: I think this should not be updatable/changeable.

Comment (kargakis) [reply pkg/api/converterfromv20180930preview.go]: Well, for starters, it needs to be set on creation. Whether it's valid or not to update should be handled in validation.

**Comment by mjudeikis** [comment]
I think its good to merge. There is a lot of mechanics here but most of it is same

File: pkg/api/converterfromv20180930preview.go (hunk 1 above)
Comment (mjudeikis) [reply pkg/api/converterfromv20180930preview.go]: true. I think we will need to shape these edge cases one by one.

**Comment by kargakis** [comment]
https://github.com/openshift/openshift-azure/issues/708

/test upgrade

**Comment by kargakis** [comment]
https://github.com/openshift/openshift-azure/issues/718

/test e2e

**Comment by kargakis** [comment]
happy upgrade test in southeastaustralia
<details>

```
$ oc logs -f e2e-azure-upgrade -c allinone
Using sync image registry.svc.ci.openshift.org/ci-op-d2kn39bq/stable:sync
Using etcdbackup image registry.svc.ci.openshift.org/ci-op-d2kn39bq/stable:etcdbackup
Using azure-controllers image registry.svc.ci.openshift.org/ci-op-d2kn39bq/stable:azure-controllers
+ [[ 1 -ne 1 ]]
+ export RESOURCEGROUP=ci-op-d2kn39bq-226a1
+ RESOURCEGROUP=ci-op-d2kn39bq-226a1
+ rm -rf _data
+ mkdir -p _data/_out
+ USE_PROD_FLAG=-use-prod=false
+ [[ -n '' ]]
+ EXEC_FLAG=
+ [[ -n '' ]]
+ UPDATE_FLAG=
+ UPDATE_EXEC_FLAG=
+ [[ -n test/manifests/normal/create.yaml ]]
+ UPDATE_FLAG=-update=test/manifests/normal/create.yaml
+ UPDATE_EXEC_FLAG='-update-exec=make e2e'
+ ARTIFACT_DIR_FLAG=
+ ARTIFACT_KUBECONFIG_FLAG=
+ [[ -n /tmp/artifacts ]]
+ ARTIFACT_DIR_FLAG=-artifact-dir=/tmp/artifacts
+ ARTIFACT_KUBECONFIG_FLAG=-artifact-kubeconfig=_data/_out/admin.kubeconfig
+ go generate ./...
bindata.go
bindata.go
+ go run cmd/createorupdate/createorupdate.go -rm -use-prod=false -update=test/manifests/normal/create.yaml -update-exec=make e2e -artifact-dir=/tmp/artifacts -artifact-kubeconfig=_data/_out/admin.kubeconfig
time="2018-11-21T10:18:56Z" level=info msg="using randomly selected region \"australiasoutheast\""
time="2018-11-21T10:18:56Z" level=info msg="creating resource group ci-op-d2kn39bq-226a1"
time="2018-11-21T10:19:01Z" level=info msg="starting the fake resource provider"
time="2018-11-21T10:19:01Z" level=info msg="generating manifest from \"test/manifests/normal/create.yaml\""
time="2018-11-21T10:19:01Z" level=info msg="starting server on localhost:8080"
time="2018-11-21T10:19:01Z" level=info msg="creating/updating cluster"
time="2018-11-21T10:19:05Z" level=info msg="convert to internal"
time="2018-11-21T10:19:05Z" level=info msg=enrich
time="2018-11-21T10:19:05Z" level=info msg="validating internal data models"
time="2018-11-21T10:19:05Z" level=info msg="generating configs"
time="2018-11-21T10:19:18Z" level=info msg="persist config"
time="2018-11-21T10:19:19Z" level=info msg="generating arm templates"
time="2018-11-21T10:19:19Z" level=info msg="write helpers"
time="2018-11-21T10:19:19Z" level=info msg="starting deploy"
time="2018-11-21T10:19:19Z" level=info msg="applying arm template deployment"
time="2018-11-21T10:19:27Z" level=info msg="waiting for arm template deployment to complete"
time="2018-11-21T10:23:31Z" level=info msg="waiting for master-000000 to be ready"
time="2018-11-21T10:24:49Z" level=info msg="waiting for master-000001 to be ready"
time="2018-11-21T10:24:50Z" level=info msg="waiting for master-000002 to be ready"
time="2018-11-21T10:24:57Z" level=info msg="waiting for infra-000000 to be ready"
time="2018-11-21T10:26:06Z" level=info msg="waiting for infra-000001 to be ready"
time="2018-11-21T10:26:15Z" level=info msg="waiting for compute-000000 to be ready"
time="2018-11-21T10:26:15Z" level=info msg="waiting for infra services to be ready"
time="2018-11-21T10:26:15Z" level=info msg="checking daemonset openshift-node/sync"
time="2018-11-21T10:26:16Z" level=info msg="checking daemonset openshift-sdn/ovs"
time="2018-11-21T10:26:16Z" level=info msg="checking daemonset openshift-sdn/sdn"
time="2018-11-21T10:26:16Z" level=info msg="checking daemonset kube-service-catalog/apiserver"
time="2018-11-21T10:26:16Z" level=info msg="checking daemonset kube-service-catalog/controller-manager"
time="2018-11-21T10:26:16Z" level=info msg="checking daemonset openshift-template-service-broker/apiserver"
time="2018-11-21T10:27:35Z" level=info msg="checking statefulset openshift-infra/bootstrap-autoapprover"
time="2018-11-21T10:27:35Z" level=info msg="checking deployment default/docker-registry"
time="2018-11-21T10:27:35Z" level=info msg="checking deployment default/router"
time="2018-11-21T10:27:36Z" level=info msg="checking deployment default/registry-console"
time="2018-11-21T10:27:37Z" level=info msg="checking deployment openshift-infra/customer-admin-controller"
time="2018-11-21T10:27:37Z" level=info msg="checking deployment openshift-ansible-service-broker/asb"
time="2018-11-21T10:27:37Z" level=info msg="checking deployment openshift-web-console/webconsole"
time="2018-11-21T10:27:37Z" level=info msg="checking deployment openshift-console/console"
time="2018-11-21T10:27:38Z" level=info msg="checking deployment openshift-monitoring/cluster-monitoring-operator"
time="2018-11-21T10:27:38Z" level=info msg="starting health check"
time="2018-11-21T10:27:38Z" level=info msg="checking console health"
time="2018-11-21T10:27:39Z" level=info msg=OK
time="2018-11-21T10:27:39Z" level=info msg="created/updated cluster"
time="2018-11-21T10:27:39Z" level=info msg="generating manifest from \"test/manifests/normal/create.yaml\""
time="2018-11-21T10:27:39Z" level=info msg="creating/updating cluster"
time="2018-11-21T10:27:39Z" level=info msg="read old config"
time="2018-11-21T10:27:39Z" level=info msg="convert to internal"
time="2018-11-21T10:27:39Z" level=info msg=enrich
time="2018-11-21T10:27:39Z" level=info msg="validating internal data models"
time="2018-11-21T10:27:39Z" level=info msg="generating configs"
time="2018-11-21T10:27:39Z" level=info msg="persist config"
time="2018-11-21T10:27:39Z" level=info msg="generating arm templates"
time="2018-11-21T10:27:39Z" level=info msg="write helpers"
time="2018-11-21T10:27:39Z" level=info msg="starting update"
time="2018-11-21T10:27:41Z" level=info msg="applying arm template deployment"
time="2018-11-21T10:27:52Z" level=info msg="waiting for arm template deployment to complete"
time="2018-11-21T10:29:01Z" level=info msg="skipping vm \"ss-master_0\" since it's already updated"
time="2018-11-21T10:29:01Z" level=info msg="skipping vm \"ss-master_1\" since it's already updated"
time="2018-11-21T10:29:01Z" level=info msg="skipping vm \"ss-master_2\" since it's already updated"
time="2018-11-21T10:29:01Z" level=info msg="skipping vm \"ss-infra_0\" since it's already updated"
time="2018-11-21T10:29:01Z" level=info msg="skipping vm \"ss-infra_1\" since it's already updated"
time="2018-11-21T10:29:02Z" level=info msg="skipping vm \"ss-compute_0\" since it's already updated"
time="2018-11-21T10:29:02Z" level=info msg="waiting for infra services to be ready"
time="2018-11-21T10:29:02Z" level=info msg="checking daemonset openshift-node/sync"
time="2018-11-21T10:29:02Z" level=info msg="checking daemonset openshift-sdn/ovs"
time="2018-11-21T10:29:02Z" level=info msg="checking daemonset openshift-sdn/sdn"
time="2018-11-21T10:29:02Z" level=info msg="checking daemonset kube-service-catalog/apiserver"
time="2018-11-21T10:29:03Z" level=info msg="checking daemonset kube-service-catalog/controller-manager"
time="2018-11-21T10:29:03Z" level=info msg="checking daemonset openshift-template-service-broker/apiserver"
time="2018-11-21T10:29:03Z" level=info msg="checking statefulset openshift-infra/bootstrap-autoapprover"
time="2018-11-21T10:29:03Z" level=info msg="checking deployment default/docker-registry"
time="2018-11-21T10:29:03Z" level=info msg="checking deployment default/router"
time="2018-11-21T10:29:04Z" level=info msg="checking deployment default/registry-console"
time="2018-11-21T10:29:04Z" level=info msg="checking deployment openshift-infra/customer-admin-controller"
time="2018-11-21T10:29:04Z" level=info msg="checking deployment openshift-ansible-service-broker/asb"
time="2018-11-21T10:29:04Z" level=info msg="checking deployment openshift-web-console/webconsole"
time="2018-11-21T10:29:05Z" level=info msg="checking deployment openshift-console/console"
time="2018-11-21T10:29:05Z" level=info msg="checking deployment openshift-monitoring/cluster-monitoring-operator"
time="2018-11-21T10:29:05Z" level=info msg="starting health check"
time="2018-11-21T10:29:05Z" level=info msg="checking console health"
time="2018-11-21T10:29:06Z" level=info msg=OK
time="2018-11-21T10:29:06Z" level=info msg="created/updated cluster"
rm -f coverage.out azure-controllers etcdbackup sync
go generate ./...
bindata.go
bindata.go
go build ./...
go build -ldflags "-X main.gitCommit=0cff7fb-dirty" ./cmd/azure-controllers
go build -ldflags "-X main.gitCommit=0cff7fb-dirty" ./cmd/etcdbackup
go build -ldflags "-X main.gitCommit=0cff7fb-dirty" ./cmd/sync

time="2018-11-21T10:29:27Z" level=info msg="deleting cluster"
time="2018-11-21T10:29:32Z" level=info msg="deleting resource group ci-op-d2kn39bq-226a1"
time="2018-11-21T10:34:14Z" level=info msg="deleted resource group ci-op-d2kn39bq-226a1"
time="2018-11-21T10:34:14Z" level=info msg="deleted cluster"
```
</details>

**Comment by kargakis** [comment]
@jim-minter ptal, also let me know your thoughts on https://github.com/openshift/openshift-azure/issues/745, I would love to get this merged asap so I won't have to rebase two different branches (this and the admin branch I am working on).

**Comment by jim-minter** [comment]
This is looking really good.  One minor side effect of the pointer removal on the internal struct is that the zero value json marshal looks like `{"plan":{},"properties":{"networkProfile":{},"authProfile":{},"servicePrincipalProfile":{},"azProfile":{}}}`, but I think that gets fixed out when we remove the json tags on internal and marshal using the external admin type.

File: pkg/api/converterfromv20180930preview.go [hunk 2]
```diff
@@ -6,92 +6,314 @@ import (
...
+
+func mergeRouterProfiles(oc *v20180930preview.OpenShiftManagedCluster, cs *OpenShiftManagedCluster) {
+	if cs.Properties.RouterProfiles == nil {
+		cs.Properties.RouterProfiles = make([]RouterProfile, 0, len(oc.Properties.RouterProfiles))
+	}
+
+	for _, rp := range oc.Properties.RouterProfiles {
+		index := -1
+		if rp.Name != nil {
+			index = routerProfileIndex(*rp.Name, cs.Properties.RouterProfiles)
```
[the remark this answers is not in the document]
Comment (kargakis) [reply pkg/api/converterfromv20180930preview.go]: This is intentional, I didn't want to bleed validation logic in the conversion function, we already added the extra layer of merging. It should also keep our validation logic on its toes.

File: pkg/api/converterfromv20180930preview.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kargakis) [reply pkg/api/converterfromv20180930preview.go]: Why can't this be part of validation?

File: pkg/api/converterfromv20180930preview_test.go [hunk 3]
```diff
@@ -4,165 +4,487 @@ import (
 	"reflect"
 	"testing"
 
+	"github.com/Azure/go-autorest/autorest/to"
```
[the remark this answers is not in the document]
Comment (kargakis) [reply pkg/api/converterfromv20180930preview_test.go]: I was planning of doing such a refactoring but got sidetracked, thanks

File: pkg/api/converterfromv20180930preview.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mjudeikis) [reply pkg/api/converterfromv20180930preview.go]: If change my provider from A to B? Old one needs to be deleted, new one added. In this case, validation might pass.

[the remark this answers is not in the document]
Comment (kargakis) [reply pkg/api/converterfromv20180930preview.go]: @mjudeikis right now we invalidate every change apart from counts. In the future we should be opening up more changes and what you bring up can be easily invalidated if we don't want to support it?

File: pkg/api/types.go [hunk 4]
```diff
@@ -8,13 +8,13 @@ const (
...
-	Tags       map[string]string     `json:"tags"`
+	Plan       ResourcePurchasePlan `json:"plan,omitempty"`
+	Properties Properties           `json:"properties,omitempty"`
+	ID         string               `json:"id,omitempty"`
+	Name       string               `json:"name,omitempty"`
+	Type       string               `json:"type,omitempty"`
+	Location   string               `json:"location,omitempty"`
+	Tags       map[string]string    `json:"tags"`
 
 	Config *Config `json:"config,omitempty"`
```
[the remark this answers is not in the document]
Comment (kargakis) [reply pkg/api/types.go]: I was touching only fields involved in the current conversions. This could be updated as part of the admin api changes? Less work for this PR which is already massive.

File: pkg/api/types.go [hunk 5]
```diff
@@ -8,13 +8,13 @@ const (
...
-	Type       string                `json:"type,omitempty"`
-	Location   string                `json:"location,omitempty"`
-	Tags       map[string]string     `json:"tags"`
+	Plan       ResourcePurchasePlan `json:"plan,omitempty"`
+	Properties Properties           `json:"properties,omitempty"`
+	ID         string               `json:"id,omitempty"`
+	Name       string               `json:"name,omitempty"`
+	Type       string               `json:"type,omitempty"`
+	Location   string               `json:"location,omitempty"`
+	Tags       map[string]string    `json:"tags"`
```
[the remark this answers is not in the document]
Comment (kargakis) [reply pkg/api/types.go]: Rest of the azure apis avoid omitempty in tags, same stands for the osa sdk we are pulling in.

File: pkg/api/converterfromv20180930preview.go [hunk 6]
```diff
@@ -6,92 +6,314 @@ import (
...
+		}
+		// If the requested profile does not exist, add it
+		// in cs as is, otherwise merge it in the existing
+		// profile.
+		if index == -1 {
+			cs.Properties.RouterProfiles = append(cs.Properties.RouterProfiles, convertRouterProfile(rp))
```
[the remark this answers is not in the document]
Comment (jim-minter) [reply pkg/api/converterfromv20180930preview.go]: Sorry.  My point is that convertRouterProfile is effectively implemented twice: once as a function, and once in the `else` case below.  I'm trying to get rid of that duplication.  When `index == -1`, you know you're going to append a new RouterProfile.  So, append it and update the index to point to it.  Then you can slot into the same codepath that updates the RouterProfile in the new new case as well.

File: pkg/api/types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jim-minter) [reply pkg/api/types.go]: Fine.

File: pkg/api/converterfromv20180930preview.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jim-minter) [reply pkg/api/converterfromv20180930preview.go]: This case can't be caught as part of a validation that takes place after a merging convert.  I think should be an explicit error (at least for now) to try to change the kind of an object via a merge.  That check could theoretically go in a separate pre-validate function, but I think it makes more sense for this function to be able to return an error.

Having made that leap, we should also insist that Name values are non-nil in the other slice merging functions as well and return an error if not.

This "pre-validation" is tightly defined because it is solely the prerequisites for the Azure-style merging to work.  Everything else belongs in .Validate().

File: pkg/api/types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jim-minter) [reply pkg/api/types.go]: True - I wonder why.  Anyway, TIL ;)

File: pkg/api/converterfromv20180930preview.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kargakis) [reply pkg/api/converterfromv20180930preview.go]: Also this is not only for empty names, but also for names not found in the existing profiles.

File: pkg/api/converterfromv20180930preview.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (kargakis) [reply pkg/api/converterfromv20180930preview.go]: @jim-minter the index gets updated on every iteration so if I update it in the loop, it's going to get overwritten during the next iteration, right? What I could do is something like the following:
```go
if index == -1 {
	cs.Properties.AgentPoolProfiles = append(cs.Properties.AgentPoolProfiles, convertAgentPoolProfile(in))
} else {
	head := append(cs.Properties.AgentPoolProfiles[:index], convertAgentPoolProfile(in))
	cs.Properties.AgentPoolProfiles = append(head, cs.Properties.AgentPoolProfiles[index+1:]...)
}
```

**Comment by kargakis** [comment]
@jim-minter comments addressed, ptal

**Comment by jim-minter** [comment]
lgtm, waiting for MSFT's OK

**Comment by jim-minter** [comment]
(rebased and force pushed)

**Comment by kargakis** [comment]
> @jim-minter your rebase dropped a bunch of work :/

Or my rebase after your rebase. Anyway, got this fixed now. I would like to merge this asap

**Comment by kargakis** [comment]
https://github.com/openshift/openshift-azure/issues/786

/retest

**Comment by kargakis** [comment]
@jim-minter I am tempted to merge, this is blocking me now.

**Comment by jim-minter** [comment]
@kargakis thanks for your patience!
